flytekit/core/shim_task.py

```diff
@@ -0,0 +1,177 @@
+      * An executor, which can use information from the task template (including the ``custom`` field)
+
+    Basically at execution time (both locally and on a Flyte cluster), the task template is given to the executor,
+    which is responsible for computing and returning the results.
+
+    .. note::
+
+       The interface at execution time will have to derived from the Flyte IDL interface, which means it may be lossy.
+       This is because when a task is serialized from Python into the ``TaskTemplate`` some information is lost because
+       Flyte IDL can't keep track of every single Python type (or Java type if writing in the Java flytekit).
+    """
+
+    def __init__(self, tt: _task_model.TaskTemplate, executor: ShimTaskExecutor, *args, **kwargs):
+        self._executor = executor
+        self._task_template = tt
+        super().__init__(*args, **kwargs)
+
+    @property
+    def task_template(self) -> _task_model.TaskTemplate:
+        return self._task_template
+
+    @property
+    def executor(self) -> ShimTaskExecutor:
+        return self._executor
+
+    def execute(self, **kwargs) -> Any:
+        """
+        Send things off to the executor instead of running here.
+        """
+        return self.executor.execute_from_model(self.task_template, **kwargs)
+
+    def dispatch_execute(
+        self, ctx: FlyteContext, input_literal_map: _literal_models.LiteralMap
+    ) -> Union[_literal_models.LiteralMap, _dynamic_job.DynamicJobSpec]:
+        """
+        Send things off to the executor instead of running here.
+        """
+        return self.executor.dispatch_execute(ctx, self.task_template, input_literal_map)
+
+
+T = TypeVar("T")
+
+
+class ShimTaskExecutor(TrackedInstance, Generic[T]):
+    @classmethod
+    def execute_from_model(cls, tt: _task_model.TaskTemplate, **kwargs) -> Any:
+        """
+        This function must be overridden and is where all the business logic for running a task should live. Keep in
+        mind that you're only working with the ``TaskTemplate``. You won't have access to any information in the task
+        that wasn't serialized into the template.
+
+        :param tt: This is the template, the serialized form of the task.
+        :param kwargs: These are the Python native input values to the task.
+        :return: Python native output values from the task.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def pre_execute(cls, user_params: ExecutionParameters) -> ExecutionParameters:
+        """
+        This function is a stub, just here to keep dispatch_execute compatibility between this class and PythonTask.
+        """
+        return user_params
+
+    @classmethod
+    def post_execute(cls, user_params: ExecutionParameters, rval: Any) -> Any:
+        """
+        This function is a stub, just here to keep dispatch_execute compatibility between this class and PythonTask.
+        """
+        return rval
+
+    @classmethod
```

Comment:
why a class method? Do we want it to be static?
I think its better to use class methods when you construct classes from it?

Reply:
nm, i see type-engine uses the same thing